File: data-structures-algorithms/leetcode/322.go
func getMinNum (num1 int, num2 int) int {
	if num1 < num2 {
		return num1
	}
	return num2
}

func coinChange(coins []int, amount int) int {
	dp := make([]int, amount+1)
	// 填充 dp
	for i := 0; i < amount + 1; i++ {
		dp[i] = amount + 1
	}
	// 当 amount 为 0 时，不需要
	dp[0] = 0
	for i := 0; i < amount + 1; i++ {
		for _, coin := range coins {
			// 金额小于零钱 那么找不开继续下一个
			if i - coin < 0 {
				continue
			}
			// 当前金额可以找开的最小个数
			dp[i] = getMinNum(dp[i], 1 + dp[i-coin])
		}
	}
	// fmt.Println(dp)
	// 找不开的情况
	if dp[amount] == amount + 1 {
		return -1
	}
    
	return dp[amount]
}

File: data-structures-algorithms/leetcode/322.js
var coinChange = function(coins, amount) {
	const rows = coins.length + 1
	const cols = amount + 1
	// 初始化
	const dp = new Array(rows).fill(0).map(() => new Array(cols).fill(cols))
	dp[0][0] = 0

	for (let row = 1; row < rows; row++) {
		dp[row][0] = 0
		for (let col = 1; col < cols; col++) {
			dp[row][col] = dp[row - 1][col]
			if (col >= coins[row - 1]) {
				dp[row][col] = Math.min(1 + dp[row][col - coins[row - 1]], dp[row][col])
			}
		}
	}

	return dp[rows - 1][cols - 1] === cols ? -1 : dp[rows - 1][cols - 1]
}